feat: compactor metrics for number of split files

carols10cents opened the pull request:
Also adds a few cleanups. I highly recommend reviewing commit-by-commit, I put in each commit message a description of why I made the changes and I tried to keep them organized.

The `feat` commit adds metrics for the number of files that need to be split in a compaction so that we can see how often files need to be split, and how many, to know whether this case needs further optimization.

Fixes https://github.com/influxdata/idpe/issues/17210.

Please let me know if I've misunderstood anything, I've been out of compactor land for a while!! 😅

Joe-Blount commented on compactor2/src/components/split_or_compact/metrics.rs:
```diff
@@ -0,0 +1,136 @@
+            .register_metric_with_options::<U64Histogram, _>(
+                METRIC_NAME_FILES_TO_SPLIT,
+                "Number of files needing to be split to minimize overlap",
+                || U64HistogramOptions::new([1, 10, 100, 1_000, 10_000, u64::MAX]),
```
If I understand correctly, this is tracking the number of files split each time we decide to do some splits.  
I was anticipating that would be a small number, so only the low end of that range would get populated.
We may also want something tracking split decisions vs compact decisions (e.g. a ratio of splits vs compacts)

@NGA-TRAN may have something else in mind though..

carols10cents replied:
Yeah, this was a total guess on my part. Please let me know what would be most useful @NGA-TRAN !

NGA-TRAN replied:
Regarding number of files to split, we want to do something similar to files to create, delete and upgrade here https://github.com/influxdata/influxdb_iox/blob/main/compactor2/src/components/commit/metrics.rs and invoked here https://github.com/influxdata/influxdb_iox/blob/309177b750c72b3f510bab5b4dc67b8491e3d7dc/compactor2/src/driver.rs#L483

I think since we already have metrics for files to create, we can use that to compare with files to split vs files to compact but I agree if we can count split decision (not number files to split) vs compact decision, we will have more metrics to analyze. This can be done as a following PR if not that simple

carols10cents replied:
Ah yes the commit metrics are exactly where I got these histogram buckets:

https://github.com/influxdata/influxdb_iox/blob/a18fe6f548e85957a74c7b87f226ab69c4a39112/compactor2/src/components/commit/metrics.rs?rgh-link-date=2023-03-08T17%3A17%3A01Z#L31

So you're saying you'd like these to stay as I have them here, right @NGA-TRAN ?

I think it'll be easy to add a counter for split/compact decision counts, let me try!

NGA-TRAN replied:
> you're saying you'd like these to stay as I have them here, right @NGA-TRAN ?

If they are the same, yes

NGA-TRAN reviewed (commented):
Your suggestion to review commit by commit makes it so easy review.

Everything looks good. I think you only want to make the metric similar to files to create, delete and upgrade, for us to put them in the dashboard easily

carols10cents commented:
> I think you only want to make the metric similar to files to create, delete and upgrade, for us to put them in the dashboard easily

Ah, I'm not sure what you mean? What part of that metric needs to be changed over here?

NGA-TRAN commented:
>> I think you only want to make the metric similar to files to create, delete and upgrade, for us to put them in the dashboard easily

> Ah, I'm not sure what you mean? What part of that metric needs to be changed over here?

Nothing needed to change. What I meant is to make the name and attributes similar to those

carols10cents commented:
> Nothing needed to change. What I meant is to make the name and attributes similar to those

I'm more confused :( Nothing needed to change? But you want me to change the name and attributes similar to those? 

Could you be more specific with what exactly you'd like me to change in this PR? Maybe with the GitHub comment suggestion feature?

carols10cents commented:
Also, I've added metrics to count the number of times split is chosen and the number of times compact is chosen, and also added more refactoring commits to reduce test duplication code, please let me know what you'd like me to change about that metric too! Thanks!

NGA-TRAN commented on compactor2/src/components/commit/metrics.rs:
```diff
@@ -362,48 +362,28 @@ mod tests {
+                        metric_name,
+                        labels = Attributes::from(&[("op", op), ("level", file_level.name())]),
+                        samples = 0,
+                        sum = 0,
```
Sorry for my ignorance but does this cover both hist_count and hist_total?

carols10cents replied:
Yup! If you take a look at the macro implementation, you'll see that `samples` is `hist_count` and `sum` is `hist_total`. This macro can do both!

NGA-TRAN reviewed (approved):
LGTM. Thanks @carols10cents
